### ISSUE TITLE
Feature request: Landing and Take Off distance

### ISSUE THREAD

**Issue by havber** [scelts/msfslandingrate#5]
Nice plugin! I would like to put in a request to also measure landing and take off distance. Would be great for STOL competitions :)

**Comment by scelts** [comment]
Ugh, this is tougher than it looks like :)
I'll take a look at it, I actually already thought about the landing distance. But I need to fix a few bugs first and sort out the code.

**Comment by havber** [comment]
Thanks for replying :) If it has the same complexity that it does in x-plane, I know that it is not straight forward. I gave that a go a while back, but quickly realised it would take more time and dedication that I was willing to offer.

**Comment by sonicviz** [comment]
I'm guessing in the landing challenges they have the gps coordinates of the target and calculate distanced from the impact point GPS coordintates to that?

**Comment by scelts** [comment]
Yeah. If there was a way to get the GPS of all the runway thresholds and directions in MSFS, it could be done as in the landing challenge.
Except for the second parameter they calculate, I have no idea what is that shit about. Sometimes I land smack on the centerline and don't move and I get low points, sometimes I just land on one side, just give up to correct it, and get high points. Have no idea what is the logic there.

Also the landing rate - still didn't figure out what they do there exactly, that's why it differs seemingly randomly from this tool.

**Comment by scelts** [comment]
I want to do this in the next bigger update if we don't end up with loads of bugs.
Can we brainstorm a bit here - I have an issue of how to define the landing/takeoff distance. Not from a perspective of implementing it, but I don't know how to define a point in landing when the landing is "done", or at the takeoff, when the takeoff has started.

The problem is of course when someone just turns to the taxiway without stopping or does a rolling takeoff.

I don't want to involve at all the runway/taxiway, even if I could read it since I want also to do this also for the bush landings.

**Comment by sonicviz** [comment]
The simplest would be to replicate the "rule" that the official landing challenges seem to use ie: you need to come to a stop before being scored.
 
Put a timer on it as well, so it's stop for at least 1 second, for example. So you need to power down, brake, stop, then wait...bingo...score. Ditto for takeoffs. only score standing starts not rolling. Rolling are a special use case anyway.

**Comment by scelts** [comment]
This is how it's done, but then this needs to be a feature you enable when you want to use, as you don't want to record this in many cases, for many users that don't do it like that.
I first, cannot remember when (even in real life), have seen people do a full stop on the active runway after landing. You'd easily get a 1.5km landing distance with a Cessna, all the way from touchdown zone to the parking :)

The best idea I have until now is to measure the distance all the time, from the point of last acceleration to the point of deceleration and vice-versa, and then to take the last when the wheels leave the ground as takeoff roll, and first after the touchdown as landing roll.

This comes with another caveat though. One can easily trick this - on a takeoff roll, you tap the breaks and continue, and you have a much shorter takeoff roll. Same on landing, with an acceleration. That's why I don't like it.
I can go down the rabbit hole, and then look for the brakes or throttle, but this is not something I like to do.

**Comment by sonicviz** [comment]
Are you talking about having it always on? I thought the use case for it was just for training, so there would be an identified start/stop behavior to trigger. 

If you want it to be "always on" then yes, it needs some heuristics to be a little smarter in order to detect when to calculate the results. But you have to wary of edge cases, such as bouncing on landing/takeoff.

**Comment by scelts** [comment]
OK, wasn't thinking about an optional feature. Will definitely consider it like that.

**Comment by sonicviz** [comment]
Well, the key here is it's for STOL competitions not general use. In that case it should be easy to constrain the start/stop and trigger conditions for measuring as it's a "drumroll" STOL competition or landing challenge you are competing in. So to do so you need to conform to the rules.  It's a training/competition aid. 

I would develop it for tightly rule constrained use cases like this first, then consider how it could be widened to a general "always on" tool that gives you general takeoff/landing feedback.

**Comment by sonicviz** [comment]
https://www.youtube.com/watch?v=zV-n9SHNZV4

**Comment by scelts** [comment]
I know this, but should this be a part of general landing monitor app, or maybe a separate, landing challenge one?
It's kind of a big change, takes a lot of code (and interface) to do, and will be used exclusively in landing challenges.

**Comment by scelts** [comment]
Will be done in some of the next updates, found a way how to do it.

**Comment by sonicviz** [comment]
Awesome. How so?

**Comment by scelts** [comment]
once you get a landing clearance from in game atc (or assign your self to an uncontrolled runway), one can read out some of the runway parameters, which you can use then to detect getting in or out from the runway. 
of course this won't work for the bush landings or when you don't use ATC, but I won't show anything then.
the requirement is that you need to use the AI ATC. which I hate for example, and it's the thing farthest from the real experience. fiddling with the menu while flying trying to find which runway is on which selection and where is what you want to say.

**Comment by sonicviz** [comment]
It seems a bit constrained (ie: no bush strips), but if you pitch it as a training tool for working on your landing skill at controlled runways (eg Sonora etc) then should be ok. Just need to be very clear about what it's for and why.

**Comment by RaffCode-Personal** [comment]
I would love to see this feature happening (I use VoiceAttack to deal with ATC at least w/o having to use M/KB by the way). Is this still in the works? 

Anyway, great addition to the sim as it is.
